week2/game_2048/class_version/game_template.py

```diff
@@ -0,0 +1,230 @@
+        for row in self.generate_index_grid(height_width):
+            self.add_merged_to_grid(row, new_grid)
+        #if self.get_grid() != new_grid:
+        #    self.set_grid(new_grid)
+        #    self.new_tile()
+        self.set_grid(new_grid)
+
+    def add_merged_to_grid(self, row, new_grid):
+        """
+        Calls merge on each row and adds it to a placeholder grid
+        """
+        temp_row = merge([self.get_tile(pair[0],pair[1]) for pair in row])
+        for index, value in enumerate(row):
+            new_grid[value[0]][value[1]] = temp_row [index]
+
+    def generate_index_grid(self, height_width):
+        """
+        Generates a grid of the indices
+        """
+        return [self.add_to_indices_grid(index, height_width) 
+                for index in self.get_move_grid_indices()[self.get_direction()]]
+
+    def add_to_indices_grid(self, index, height_width):
+        """
+        Adds each row of indices to the grid
+        """
+        return [index] + self.generate_row_indices(index[0], index[1], height_width)
+
+    def generate_row_indices(self, index0, index1, height_width):
+        """
+        Generates each row of indices to insert into the grid
+        """
+        temp = []
+        for dummy_number in range(height_width - 1):
+            index0 += OFFSETS[self.get_direction()][0]
+            index1 += OFFSETS[self.get_direction()][1]
+            temp.append((index0, index1))
+        return temp
+
+    def new_tile(self):
+        """
+        Create a new tile in a randomly selected empty
+        square.  The tile should be 2 90% of the time and
+        4 10% of the time.
+        """
+        row, column = self.select_tile()
+        self.set_tile(row, column, 2) if random.randint(0,100) < 90 else self.set_tile(row, column, 4)
+
+    def check_tile_is_empty(self, row, col):
+        """
+        Takes in a row and column and returns True
+        if the value of grid[row, column] == 0  
+        """
+        return self.get_tile(row, col) == 0
+
+    def select_tile(self):
+        """
+        Returns a tuple with a randomly selected coordinate 
+        pair in the grid that has a current value of 0
+        """
+        return self.get_empty_tile(self.get_random_row_index(), self.get_random_col_index())
+
+    def get_empty_tile(self, row, column):
+        """
+        Takes in the indexes of a tile in the grid and returns
+        it if it has a value of 0, if not it returns the 
+        indexes of a tile that does have a value of 0.
+        """
+        if not self.check_tile_is_empty(row, column):
+            return self.get_empty_tile(self.get_random_row_index(), self.get_random_col_index())
+        return row, column
+
+    def get_random_row_index(self):
+        """
+        Returns a random index between 0 and get_grid_height
+        """
+        return random.randint(0, self.get_grid_height() - 1)
+
+    def get_random_col_index(self):
```

Comment:
Should I combine these and just pass in a parameter for the range? I was that separating them would increase readability and decrease the chance of passing in the wrong size to range... but I am not sure that it does either of those effectively. 
```
    def get_random_row_index(self):
        """
        Returns a random index between 0 and get_grid_height
        """
        return random.randint(0, self.get_grid_height() - 1)

    def get_random_col_index(self):
        """
        Returns a random index between 0 and get_grid_width
        """
        return  random.randint(0, self.get_grid_width() - 1)
```

Reply:
If the range is a property of the class you've done this correctly. Alternatively, if it isn't an inherent property of the class then you can make a helper function (class function) that calculates a random number in a range.